src/pseudopeople/noise_functions.py

```diff
@@ -149,13 +153,144 @@ def duplicate_with_guardian(
+    Function that duplicates rows of a dataset. Rows that are duplicated fall into one of three groups of
+    dependents that are typically duplicated in administrative data where a dependent lives in a different
+    location then their guardian and they show up in the data at both locations. For simplicity, rows will
+    be only duplicated once for a maximum of two rows per dependent. When a row is duplicated, one row will
+    have the depdent's correct address and the other will have the guardian's address.
+    :param dataset_name: Name of the dataset being noised
+    :param dataset_data: pd.DataFrame that will be noised
+    :param configuration: ConfigTree object containing noise level values. Dict with 3 key groups for duplication
+    :param randomness_stream: RandomnessStream instance to make random selection for noise
     :return:
     """
+
+    # dataset_data = dataset_data.copy()
+    # Get dict of group type and formatted dataframe for that group that should be noised
+    formatted_group_data = {}
+    # Get dataframe for each dependent group to merge with guardians
+    in_households_under_18 = dataset_data.loc[
+        (dataset_data["age"] < 18)
+        & (dataset_data["housing_type"] == "Household")
+        & (dataset_data["guardian_1"].notna())
+    ]
+    in_households_18_to_23 = dataset_data.loc[
+        (dataset_data["age"] >= 18)
+        & (dataset_data["age"] < 24)
+        & (dataset_data["housing_type"] == "Household")
+        & (dataset_data["guardian_1"].notna())
+    ]
+    in_group_quarters_under_24 = dataset_data.loc[
+        (dataset_data["age"] < 24)
+        & (dataset_data["housing_type"] != "Household")
+        & (dataset_data["guardian_1"].notna())
+    ]
+
+    # Helper function to format group dataframe and merging with their dependentss
+    def _merge_depdents_and_guardians(
+        dependents_df: pd.DataFrame, full_data: pd.DataFrame
+    ) -> pd.DataFrame:
+        # Merge dependents with their guardians. We have to merge twice to check
+        # if either guardian is living at a separate location from the dependent.
+        guardian_1s = full_data.loc[
+            full_data["simulant_id"].isin(full_data["guardian_1"])
+        ].add_prefix("guardian_1_")
+        dependents_and_guardians_df = dependents_df.merge(
+            guardian_1s,
+            how="left",
+            left_on=["guardian_1", "year"],
+            right_on=["guardian_1_simulant_id", "guardian_1_year"],
+        )
+        del guardian_1s
+        guardian_2s = full_data.loc[
+            full_data["simulant_id"].isin(full_data["guardian_2"])
+        ].add_prefix("guardian_2_")
+        dependents_and_guardians_df = dependents_and_guardians_df.merge(
+            guardian_2s,
+            how="left",
+            left_on=["guardian_2", "year"],
+            right_on=["guardian_2_simulant_id", "guardian_2_year"],
+        )
+        del guardian_2s
+        return dependents_and_guardians_df
```

Comment:
What exactly happens when a guardian themselves has a guardian? Is that even a possibility? If so, does the lowest-level dependent get the copied values from their guardian?

Reply:
Yes that happens in the test.  Simulant id 2 has guardian who is simulant id 5 and simulant id 5 has guardian of simulant id 6. I haven't looked in the actual data to see if that happens but it seems feasible so I tested that.

Reply:
I feel like this situation needs to be clarified w/ RT. Like, if simulant_1's guardian is simulant_2 and simulant_2's guardian is simulant_3, and both simulant_1 and _2 copy from guardian, would simulant_1 copy simulant_2 before or after simulant_2 copies simualnt_3?

Reply:
They would copy from their guardian.  There is no "order" problem here because we merge guardians onto their dependents. We aren't copying row by row.